### PR TITLE
Paul/paybis

### PR DESCRIPTION
**Pull request by paullinator**
### CHANGELOG

 none

### Dependencies

 none

### Requirements

If you have made **any** visual changes to the GUI. Make sure you have:

- [ ] Tested on iOS device
- [ ] Tested on Android device
- [ ] Tested on small-screen device (iPod Touch)
- [ ] Tested on large-screen device (tablet)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205041005078292

### CODE REVIEW

File: src/plugins/gui/providers/paybisProvider.ts [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+          displayCurrencyCode,
+          direction,
+          tokenId
+        } = params
+        const paymentType = paymentTypes.find(paymentType => allowedPaymentTypes[direction][paymentType] === true)
+        if (paymentType == null) throw new FiatProviderError({ providerId, errorType: 'paymentUnsupported' })
+
+        const pairs = paybisPairs?.data
+        if (pairs == null) {
+          throw new FiatProviderError({ providerId, errorType: 'assetUnsupported' })
```
Comment (Jon-edge) [inline src/plugins/gui/providers/paybisProvider.ts:302]: Seems a odd to be throwing this without comparing to the getQuote params

File: src/plugins/gui/providers/paybisProvider.ts [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+              throw new FiatProviderError({ providerId, errorType: 'underLimit', errorAmount: Number(minMatch[1]), displayCurrencyCode: minMatch[2] })
+            }
+          }
+          throw new Error(lastError?.error.message ?? 'Paybis Unknown paymentMethodError')
+        }
+
+        // Should only have one payment method
+        if (paymentMethods.length !== 1) {
+          throw new Error('Invalid number of quoted payment methods')
+        }
```
Comment (Jon-edge) [inline src/plugins/gui/providers/paybisProvider.ts:368]:
```suggestion
        assert(paymentMethods.length !== 1, 'Invalid number of quoted payment methods')
```

File: src/plugins/gui/providers/paybisProvider.ts [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+            const { coreWallet, showUi } = approveParams
+            const receiveAddress = await coreWallet.getReceiveAddress()
+
+            const bodyParams = {
+              cryptoWalletAddress: {
+                currencyCode: paybisCc,
+                address: receiveAddress.segwitAddress ?? receiveAddress.publicAddress
+              },
+              partnerUserId,
+              locale: locale.localeIdentifier.slice(0, 2),
```
Comment (Jon-edge) [inline src/plugins/gui/providers/paybisProvider.ts:412]: We should start referencing the same locale logic i.e. pickLanguage() in
preparation for locale UI settings

File: src/plugins/gui/pluginUtils.ts [hunk 4]
```diff
@@ -128,3 +128,16 @@ export const debugSpewStore = async (store: EdgeDataStore): Promise<void> => {
   console.log('*** theStore ***')
   console.log(JSON.stringify(theStore, null, 2))
 }
+
+export const assert = (condition: boolean, message: string) => {
+  if (!condition) throw new Error(message)
+}
```
Comment (Jon-edge) [inline src/plugins/gui/pluginUtils.ts:134]: Might not be worth it, but we can also have a generic flavor:

```suggestion
export const assert = (
  condition: boolean,
  ErrorType: new (...args: any[]) => Error = Error,
  props: Record<string, any> = { message: 'An error occurred' }
) => {
  if (!condition) {
    const error = new ErrorType()
    Object.assign(error, props)
    throw error
  }
}

// usage:
assert(true, undefined, { message: "Something went wrong" })
assert(minMatch != null, FiatProviderError, { providerId, errorType: 'underLimit', errorAmount: Number(minMatch[1]), displayCurrencyCode: minMatch[2] })

```

File: src/plugins/gui/providers/paybisProvider.ts (hunk 1 above)
Comment (paullinator) [reply src/plugins/gui/providers/paybisProvider.ts:302]: This error implies we were not able to get the supported assets in the first place.

File: src/plugins/gui/providers/paybisProvider.ts (hunk 2 above)
Comment (paullinator) [reply src/plugins/gui/providers/paybisProvider.ts:368]: Nice. I'll use it.

File: src/plugins/gui/pluginUtils.ts (hunk 4 above)
Comment (paullinator) [reply src/plugins/gui/pluginUtils.ts:134]: I like it, but I don't want to add the code now as this PR won't be using it. I'm open to it being added once it once it's used.

File: src/plugins/gui/providers/paybisProvider.ts (hunk 1 above)
Comment (Jon-edge) [reply src/plugins/gui/providers/paybisProvider.ts:302]: Ah right this was the error type that was used as more like a result

**Review by Jon-edge — APPROVED** [review]

**Pull request merged**